cmd/debugger/main.go

```diff
@@ -1,58 +1,129 @@
 package main
 
 import (
+	"context"
+	"encoding/json"
 	"fmt"
 	"io"
 	"io/ioutil"
 	"net"
 	"os"
 	"os/exec"
+	"time"
+
+	"github.com/earthly/earthly/debugger/common"
+
+	"github.com/creack/pty"
+	"github.com/hashicorp/yamux"
 )
 
 var (
 	// Version is the version of the debugger
 	Version string
+
+	// ErrNoShellFound occurs when the container has no shell
+	ErrNoShellFound = fmt.Errorf("no shell found")
 )
 
-func interactiveMode(remoteConsoleAddr string) {
-	c, err := net.Dial("tcp", remoteConsoleAddr)
+func getShellPath() (string, bool) {
+	for _, sh := range []string{
+		"bash", "ksh", "zsh", "sh",
+	} {
+		if path, err := exec.LookPath(sh); err == nil {
+			return path, true
+		}
+	}
+	return "", false
+}
+
+func interactiveMode(remoteConsoleAddr string) error {
+	conn, err := net.Dial("tcp", remoteConsoleAddr)
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "failed to connect to earth remote console: %v\n", err)
-		return
+		return err
 	}
+	defer func() {
+		err := conn.Close()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "error closing: %v\n", err)
+		}
+	}()
 
-	cmd := exec.Command("sh", "-i")
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
+	session, err := yamux.Client(conn, nil)
+	if err != nil {
+		return err
+	}
 
-	stdin, err := cmd.StdinPipe()
+	ptyStream, err := session.Open()
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "failed to attach to shell stdin: %v\n", err)
-		return
+		return err
 	}
+	ptyStream.Write([]byte{common.PtyStream})
 
-	err = cmd.Start()
+	winChangeStream, err := session.Open()
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "failed to exec debugger shell: %v\n", err)
-		return
+		return err
 	}
+	winChangeStream.Write([]byte{common.WinChangeStream})
 
-	b := make([]byte, 256)
-	for {
-		n, err := c.Read(b)
-		if err != nil {
-			if err != io.EOF {
-				fmt.Printf("err: %v\n", err)
-			}
-			break
-		}
-		stdin.Write(b[:n])
+	shellPath, ok := getShellPath()
+	if !ok {
+		return ErrNoShellFound
 	}
-	err = cmd.Wait()
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "failed to wait for debugger shell: %v\n", err)
-		return
+	c := exec.Command(shellPath)
```

Comment:
Should we pass in a context here too, for cancellation purposes? (Just in case shell hangs)